tests/test_pull.py

```diff
@@ -94,13 +119,29 @@ def test_remove_deprecated_images(run):
     pull.remove_deprecated_images(local_images)
 
 
+@pytest.mark.parametrize("project_yaml,exc_msg", [
+    ("doesnotexist.yaml", "Could not find"),
+    ("bad.yaml", "Could not parse"),
+    ("noactions.yaml", "No actions found"),
+])
```

Comment:
This is really nice 👍

Reply:
Thanks. I'm a big fan of ~2 pytest tests for a function. 1 to test the happy path, and a parameterised one that tests the error paths, which should all raise an exception of some kind. I like partial exc message matching too, as it reduces test brittleness.

Reply:
/me side eyes the many failure tests for job-sever

🤔

Reply:
Sometimes its not always doable, but I've found it a good default for smaller functions.

For more complex cases, I use a generator function to generate the test cases.